speaker/layer2_controller.go

```diff
@@ -97,15 +115,42 @@ func (c *layer2Controller) ShouldAnnounce(l log.Logger, name string, svc *v1.Ser
 }
 
 func (c *layer2Controller) SetBalancer(l log.Logger, name string, lbIP net.IP, pool *config.Pool) error {
+	// Determine if the list of active Nodes as determined by memberlist has
+	// changed since the last time we checked this Service's IP.  If so, we kick off
+	// sending gratuitous ARP / NDP for this Service IP.  This will handle the
+	// case where the cluster membership change is recovering from a network
+	// partition and we need to establish who the real current owner is of each
+	// IP address.  This happens asynchronously.
+	var activeNodes map[string]bool
+	var newService bool
+	activeNodes = c.getActiveNodes()
+	if _, ok := lastActiveNodes[lbIP.String()]; !ok {
+		// If it wasn't in the lastActiveNodes map, we hadn't seen this Service yet,
+		// so we shouldn't trigger try triggering re-announcement.
+		newService = true
+		lastActiveNodes[lbIP.String()] = activeNodes
+	}
+	if !newService && !reflect.DeepEqual(activeNodes, lastActiveNodes[lbIP.String()]) {
+		l.Log("op", "SetBalancer", "msg", "Detected cluster membership change, re-announcing IP",
+			"newActiveNodes", activeNodes, "oldActiveNodes", lastActiveNodes[lbIP.String()])
+		c.announcer.ReAnnounce(name)
+		lastActiveNodes[lbIP.String()] = activeNodes
+	}
+
 	c.announcer.SetBalancer(name, lbIP)
 	return nil
 }
 
 func (c *layer2Controller) DeleteBalancer(l log.Logger, name, reason string) error {
 	if !c.announcer.AnnounceName(name) {
+		// No longer tracked by the announcer, so nothing else to do.
 		return nil
 	}
-	c.announcer.DeleteBalancer(name)
+	var unusedIP net.IP
+	unusedIP = c.announcer.DeleteBalancer(name)
+	if unusedIP != nil {
+		delete(lastActiveNodes, unusedIP.String())
```

Comment:
I'm a bit confused by this. In which cases exactly does `announcer.DeleteBalancer()` return an IP, and which IP is it?